### PR TITLE
[P2P / API] Add ability to ban and un-ban peer from API call

### PR DESCRIPTION
**Pull request by fassadlr**

### CODE REVIEW

**Comment by fassadlr** [comment]
@bokobza This implementation was based of my own design. It appears that Bitcoin Core doe not implement an API call to disconnect peers.

https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_calls_list

@maciejzaleski requested that we implement a mechanism whereby we can disconnect and ban misbehaving peers via an API call. @dangershony and I discussed on Friday and we decided to implement it in this manner.

I assumed that the `NodeController` is the correct place for this as we are accessing FullNode features here. I'm not sure that this is RPC specific? 

WDYT?

**Comment by bokobza** [comment]
There are RPC methods for all of those.
Check out https://bitcoin.org/en/developer-reference#network-rpcs
You have `clearbanned`, `disconnectnode`, `setban` etc...
You can do it via the Api but then maybe a new `/network` path and controller will be more appropriate. And you might as well add that to the RPC commands.

File: src/Stratis.Bitcoin/Connection/PeerBanning.cs [hunk 1]
```diff
@@ -126,5 +132,23 @@ public bool IsBanned(IPEndPoint endpoint)
...
+            List<PeerAddress> peerAddresses = this.peerAddressManager.FindPeersByIp(endpoint);
+            if (peerAddresses.Count == 0)
+            {
+                this.logger.LogTrace("(-)[NO_PEERS_TO_UNBAN]");
+                return;
+            }
+
+            foreach (var peerAddress in peerAddresses)
+            {
+                peerAddress.UnBan();
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Connection/PeerBanning.cs]: Will this also update the store?

File: src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+namespace Stratis.Bitcoin.Controllers.Models
+{
+    /// <summary>
+    /// Represents the model that will ban and disconnect a connected peer.
+    /// </summary>
+    public sealed class BanPeerViewModel
+    {
+        public BanPeerViewModel()
+        {
+            this.BanDurationSeconds = TimeSpan.FromDays(1).Seconds;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs]: Should we default this?

**Comment by fassadlr** [comment]
@bokobza please confirm the following changes to the PR:

1. Create `NetworkController` with path `\network`
2. Move the existing ban and un-ban code there but change so that we have `disconnect`, `setban` and `clearbanned` methods in the API. 
3. Add to the RPC controller the following:
-disconnect
-setban
-clearbanned

File: src/Stratis.Bitcoin/Connection/PeerBanning.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Connection/PeerBanning.cs]: Yes it does, this is tested in a set of new integration tests.

File: src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs]: `IPeerBanning.BanAndDisconnectPeer` requires a value to be passed, so defaulting it here in the constructor is OK?

Comment (dangershony) [reply src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs]: But the default should be taken from the settings

**Comment by fassadlr** [comment]
image

The new controller methods in swagger.

File: src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Controllers/Models/BanPeerViewModel.cs]: Oh yes, my bad, I see it now in `Settings`, OK I will implement it now 👍

File: src/Stratis.Bitcoin/Controllers/NetworkController.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        /// </summary>
+        /// <param name="viewModel">The model that represents the peer to disconnect.</param>
+        /// <returns><see cref="OkResult"/></returns>
+        [Route("disconnect")]
+        [HttpPost]
+        public IActionResult DisconnectPeer([FromBody] DisconnectPeerViewModel viewModel)
+        {
+            try
+            {
+                var endpoint = viewModel.PeerAddress.ToIPEndPoint(0);
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Controllers/NetworkController.cs]: I think this should be

```
    var endpoint = viewModel.PeerAddress.ToIPEndPoint(nodeSettings.Network.DefaultPort)
```

Comment (dangershony) [reply src/Stratis.Bitcoin/Controllers/NetworkController.cs]: In fact should we not actually pass the entire end point not just the address?
Currently we cannot disconnect peers that are on a different port

File: src/Stratis.Bitcoin/Controllers/NetworkController.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        /// </summary>
+        /// <param name="viewModel">The model that represents the peer to add or remove from the banned list.</param>
+        /// <returns><see cref="OkResult"/></returns>
+        [Route("setban")]
+        [HttpPost]
+        public IActionResult SetBan([FromBody] SetBanPeerViewModel viewModel)
+        {
+            try
+            {
+                var endpoint = new IPEndPoint(IPAddress.Parse(viewModel.PeerAddress), 0);
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Controllers/NetworkController.cs]: Why not use the `ToIPEndPoint` here as well?

File: src/Stratis.Bitcoin/Controllers/NodeController.cs [hunk 5]
```diff
@@ -69,14 +69,20 @@ public class NodeController : Controller
         /// <summary>An interface implementation for the blockstore.</summary>
         private readonly IBlockStore blockStore;
 
-        public NodeController(IFullNode fullNode, ILoggerFactory loggerFactory,
-            IDateTimeProvider dateTimeProvider, IChainState chainState,
-            NodeSettings nodeSettings, IConnectionManager connectionManager,
-            ConcurrentChain chain, Network network, IPooledTransaction pooledTransaction = null,
-            IPooledGetUnspentTransaction pooledGetUnspentTransaction = null,
+        public NodeController(
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Controllers/NodeController.cs]: Whats all the changes in this class? just a reorder?

File: src/Stratis.Bitcoin/Controllers/NetworkController.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Controllers/NetworkController.cs]: Fixed 👍

File: src/Stratis.Bitcoin/Controllers/NodeController.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Controllers/NodeController.cs]: Yeah just to reorder 👍

File: src/Stratis.Bitcoin/Controllers/NetworkController.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/Controllers/NetworkController.cs]: Fixed 👍

**Comment by dangershony** [comment]
@maciejzaleski actually does not need the `forcelisten` flag so I asked him to remove it.

**Pull request merged**